zip-0308.rst

```diff
@@ -0,0 +1,326 @@
+be interpreted as described in RFC 2119. [#RFC2119]_
+
+The terms below are to be interpreted as follows:
+
+Sprout protocol
+  Code-name for the Zcash shielded protocol at launch.
+Sapling protocol
+  Code-name for the Zcash shielded protocol added by the second Zcash network
+  upgrade, also known as Network Upgrade 1.
+
+
+Abstract
+========
+
+This proposal describes privacy-preserving procedures to migrate funds from
+Sprout to Sapling z-addresses; and supporting RPC operations to enable,
+disable, and monitor the migration process.
+
+
+Motivation
+==========
+
+Zcash Sapling [#zip-0205]_ introduces significant efficiency improvements
+relative to the previous iteration of the Zcash shielded protocol, Sprout.
+These improvements will pave the way for broad mobile, exchange and vendor
+adoption of shielded addresses.
+
+Therefore, we anticipate that users will want to migrate all their shielded
+funds from Sprout to Sapling.
+
+The Zcash consensus rules prohibit direct transfers from Sprout to Sapling
+z-addresses, unless the amount is revealed by sending it through the
+"transparent value pool" [#transparent-value-pool]_. The primary motivation
+for this is to allow detection of any overall inflation of the Zcash monetary
+base, due to exploitation of possible vulnerabilities in the shielded
+protocols or their implementation.
+
+Since the exposure of the migrated amount potentially compromises the privacy
+of users, we wish to define a way to perform the migration that mitigates
+this privacy leak as far as possible. This can be done by hiding individual
+migration transactions among those of all users that are doing the migration
+at around the same time.
+
+The security analysis of migration strategies is quite subtle; the more
+obvious potential strategies can leak a lot of information.
+
+
+Requirements
+============
+
+Migration is performed "in the background" by a ``zcashd`` node. It does not
+significantly interfere with concurrent usage of the node, other than possibly
+increasing the latency of some other shielded operations.
+
+It is possible to enable or disable migration at any time.
+
+All shielded funds in Sprout z-addresses will eventually be transferred to
+Sapling z-addresses, provided the node is working.
+
+It should take a "reasonable" length of time to complete the transfer;
+typically less than 2 months.
+
+The design should mitigate information leakage via timing information and
+transaction data, including
+
+* linkage of particular z-addresses or users, and the amounts held;
+* information about the distribution of amounts of individual notes.
+
+The design and implementation is stateless, to the extent practical.
+
+Visibility is provided for the wallet/node user into the progress of the
+migration.
+
+There is sufficient information available to debug failed transactions that
+are part of the migration.
+
+The design recovers from failed operations to the extent possible.
```

Comment:
How does it recover?  Just  rerun the operation based on the current remaining balance to migrate?

Reply:
The wallet doesn't keep much state (apart from whether the migration is enabled), so it automatically recovers on the next batch.